### ISSUE TITLE
Charging Lighting (Feature Request)

### ISSUE THREAD

**Issue by Decapitater** [polychromatic/polychromatic#371]
Hi , 
can charging lighting be added for all the supported mouse with docks ?
I quote the synapse 
"" Red < 25 %  , Orange 26-50 % , Yellow %1-75% , Green (Breathing) 76-99% , Green (static) 100 %""
At the current status , only the mouse shows the indication and the dock remains the decided color.
This also creates the async bug with causes the mouse to async from the rest of the equipments.

(Note for devs : i really appreciate the work that you guys provide , i have razer hardware and would like to assist in production of this awesome project)

**Comment by lah7** [comment]
Sorry for the slight confusion in the original issue in #369. At least now we understand a bit more of what's happening - https://github.com/openrazer/openrazer/issues/1613#issuecomment-956119284

This would be a custom effect, but it'll be a while before this gets implemented, as I need to add the "scripted effect" pieces into place. [On the roadmap](https://docs.polychromatic.app/roadmap/), I'm looking to have it completed for v1.0.0.

---

Since I haven't seen this effect before - this is how I understand the current situation:

- The mouse always overrides the effect when docked.
- The mouse doesn't go back to its previous settings when undocked.
- Under Synapse, the dock hardware usually changes the effect to match the mouse's battery %, but currently does nothing.

Please correct me if that's not right.

#### What we want:

- Support the ability to set effects when docked / undocked, (or most likely, charging starts / stops).
- Replicate the software charging effect, a bonus if it can be customised.

#### OpenRazer developments desired:

- Fix the battery status bug when docked wirelessly - https://github.com/openrazer/openrazer/issues/1613#issuecomment-956119284
- Add an API to retrieve the docked state (if any).
- Emit events when charging starts/stops and docked/undocked (if possible).

So, in a nutshell, some improvements in OpenRazer will make this feasible for future use. Currently, this can only be possible in wired mode.

**Comment by Decapitater** [comment]
-Yes it does 
- The mouse does go back to its previous settings however it does not sync with the other its 1-2 sec's behind but it shows the same spectrum
- yes , in synapse both the mouse and dock shows the battery respective charging status & returns to the synced spectrum. 
 In our case the mouse changes but the dock does not change also the mouse gets asynced 
- Yes + retains the choosed effect when undocked/does not async 
- yes
- yes the charging does not show and its useless in wired case when directly connected with wire
- yes

**Comment by Decapitater** [comment]
@lah7  @z3ntu  The pcap file is ready

**Comment by Decapitater** [comment]
its been 4 days since trying to reach

**Comment by lah7** [comment]
Please be patient. I had a look at the pcaps myself out of curiosity but I'm not knowledgeable on what to do further. It'll be looked at by OpenRazer, but it's not urgent. We're all busy. 

Should note that everytime you send multiple comments, it sends a 🔔 /📧 to many people, so please be courteous about that.

**Comment by Decapitater** [comment]
Sure @lah7  if i provide the analysis , how would it effect the development

**Comment by lah7** [comment]
Your part is done, thanks. Now we just wait for the OpenRazer maintainer to have a look - who's currently **sick** 🤢 - so it could be a while yet. **z3ntu** possesses the driver/pcaps knowledge, I work on this project (a frontend). I left my findings here: https://github.com/openrazer/openrazer/issues/1613#issuecomment-962627472

I **do not** usually work (or fully understand) pcap/driver stuff so my comments was to [help have them ready](https://github.com/openrazer/openrazer/issues/1613#issuecomment-955183236) for later.

Development can have its slow times in the open source world, but pestering for an update/response doesn't help. For sure, for this feature request, it may be another month / next year until Polychromatic is ready for this software effect, due to other priorities that need to happen first.

**Comment by Decapitater** [comment]
@lah7  btw who is maintainer of this repo

**Comment by lah7** [comment]
I maintain this repo (Polychromatic), this is my project I work on. I co-maintain OpenRazer too, usually helping with support tickets and fixing the occasional Python bug, but I consider @z3ntu the primary maintainer who knows more about Razer device internals and specifications.

**Comment by Decapitater** [comment]
Not a feature request anymore ?

**Comment by lah7** [comment]
I was renaming a few recent `feature request` issues since it's redundant to also have "Feature Request" in the title. The tag remains, nothing has changed here.

Just noticed a similar, older request for power triggers (#164), which would complement this custom effect (#371).

**Comment by Bitmods** [comment]
I was looking for a similar feature, and noticed several others posts trying to monitor their battery % of Razer mouse+docks. I wrote this script which works perfectly for me using polychromatic-cli and cron scheduling: https://github.com/Bitmods/razerbatterymonitor

Hopefully it gets officially added to polychromatic soon as its such a good feature!

**Comment by Decapitater** [comment]
@lah7 want to be the first to try ?

**Comment by HanM23** [comment]
> -Yes it does
> 
>     * The mouse does go back to its previous settings however it does not sync with the other its 1-2 sec's behind but it shows the same spectrum
> 
>     * yes , in synapse both the mouse and dock shows the battery respective charging status & returns to the synced spectrum.
>       In our case the mouse changes but the dock does not change also the mouse gets asynced
> 
>     * Yes + retains the choosed effect when undocked/does not async
> 
>     * yes
> 
>     * yes the charging does not show and its useless in wired case when directly connected with wire
> 
>     * yes

May i add something, because i think this is related.

I have the same behavior as @Decapitater, so i would be very happy to have those improvements. 
In addition, when i dock my Viper Ultimate on the dock, the device "Razer Viper Ultimate (Wired) does not appear within Polychromatic, i only have both devices : "Razer Mouse dock" and "Razer Viper Ultimate (Wireless)".

I also want to **thank all of you** for Polychromatic developments, it's a wonderful software so far.

Regards,

**Comment by lah7** [comment]
@Bitmods Thanks for sharing your solution, I like the detailed colour gradient based on percentage! 👍🏼 

@HanM23 Glad you like it! I left a comment on https://github.com/openrazer/openrazer/issues/1613#issuecomment-986236580, as I believe this is expected behaviour. It seems like the mouse remains in "wireless" mode while docked. Maybe this protocol makes sense for computers, but it seems silly for us humans. 😀 

---

Just to clarify: there are two parts to this issue. I am slowly but surely working towards them:

- **Scripted Effects:** A new format for software effects written in Python. This will be a new selectable effect for monitoring battery % and changing colours accordingly using a new API. This doesn't depend on any OpenRazer changes.

    This will be more efficient (CPU wise) then summoning cron/bash/CLI every minute, although @Bitmods's solution works reliably  too and is a good solution for the time being.

- **Triggers:** Conditionally (and optionally) switch effects upon conditions, like when charging ([#164](https://github.com/polychromatic/polychromatic/issues/164)) or time of day ([#272](https://github.com/polychromatic/polychromatic/issues/272)). This is where some work is needed in OpenRazer to support the dock state, if possible.

**Comment by Decapitater** [comment]
@lah7  thanks for the much needed explanation

**Comment by HanM23** [comment]
@lah7 I associate myself to @Decapitater to thank you for the clarification
